src/pudl/helpers.py

```diff
@@ -332,90 +332,271 @@ def is_doi(doi):
-    if right_on not in right.columns:
-        raise ValueError(f"Right dataframe has no {right_on}.")
-    missing_left_cols = [col for col in by if col not in left.columns]
-    if missing_left_cols:
-        raise ValueError(f"Left dataframe is missing {missing_left_cols}.")
-    missing_right_cols = [col for col in by if col not in right.columns]
-    if missing_right_cols:
-        raise ValueError(f"Left dataframe is missing {missing_right_cols}.")
-
-    def cleanup(df, on, by):
-        df = apply_pudl_dtypes(df)
-        df.loc[:, on] = pd.to_datetime(df[on])
-        df = df.sort_values([on] + by)
+
+    def separate_date_cols(df, date_col_name, date_on):
+        df[date_col_name] = pd.to_datetime(df[date_col_name])
+        if "year_temp_for_merge" in date_on:
+            df.loc[:, "year_temp_for_merge"] = df[date_col_name].dt.year
+        if "quarter_temp_for_merge" in date_on:
+            df.loc[:, "quarter_temp_for_merge"] = df[date_col_name].dt.quarter
+        if "month_temp_for_merge" in date_on:
+            df.loc[:, "month_temp_for_merge"] = df[date_col_name].dt.month
+        if "day_temp_for_merge" in date_on:
+            df.loc[:, "day_temp_for_merge"] = df[date_col_name].dt.day
         return df
 
-    return pd.merge_asof(
-        cleanup(df=left, on=left_on, by=by),
-        cleanup(df=right, on=right_on, by=by),
-        left_on=left_on,
-        right_on=right_on,
-        by=by,
-        tolerance=pd.Timedelta("365 days"),  # Should never match across years.
+    right = convert_col_to_datetime(right, right_date_col)
+    left = convert_col_to_datetime(left, left_date_col)
+    date_on = _add_suffix_to_date_on(date_on)
+    right = separate_date_cols(right, right_date_col, date_on)
+    left = separate_date_cols(left, left_date_col, date_on)
+    merge_cols = date_on + on
+    out = pd.merge(left, right, on=merge_cols, how=how, **kwargs)
+
+    suffixes = ["", ""]
+    if left_date_col == right_date_col:
+        if "suffixes" in kwargs:
+            suffixes = kwargs["suffixes"]
+        else:
+            suffixes = ["_x", "_y"]
+    # reconstruct the new report date column and clean up columns
+    left_right_date_col = [left_date_col + suffixes[0], right_date_col + suffixes[1]]
+    if report_at_start:
+        # keep the later of the two report dates when determining
+        # the new report date for each row
+        reconstructed_date = out[left_right_date_col].max(axis=1)
+    else:
+        # keep the earlier of the two report dates
+        reconstructed_date = out[left_right_date_col].min(axis=1)
+    out = out.drop(left_right_date_col + date_on, axis=1)
+    out.insert(loc=0, column=new_date_col, value=reconstructed_date)
+    return out
+
+
+def expand_timeseries(
+    df: pd.DataFrame,
+    key_cols: List[str],
+    date_col: str = "report_date",
+    freq: str = "MS",
+    fill_through_freq: Literal["year", "month", "day"] = "year",
```

Comment:
This argument was added so now you can fill through the end of a desired time frequency, not just year. I think this was what was being discussed [in this thread](https://github.com/catalyst-cooperative/pudl/pull/1550#discussion_r883931974). This makes the whole thing more confusing and less clean so if this functionality doesn't seem necessary maybe it should get taken out. I also added a test for filling through the end of the month.